platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -1078,28 +1081,14 @@ public void removeOutgoingEdges() {
             return;
         }
 
-        boolean alreadyRemoving = removingOutgoingEdges;
-        removingOutgoingEdges = true;
-        if (!outgoingEdges.isEmpty() && !alreadyRemoving) {
+        if (!outgoingEdges.isEmpty()) {
             for (EdgeState outgoingEdge : outgoingEdges) {
-                outgoingEdge.markUnused(); // Track that these edges have been removed from outgoing but maybe not from incoming, in case we hit one of these continues and do not call disconnectOutgoingEdge
-                ComponentState targetComponent = outgoingEdge.getTargetComponent();
-                if (targetComponent == component) {
-                    // if the same component depends on itself: do not attempt to cleanup the same thing several times
-                    continue;
```

Comment:
This continue was the original cause of the bug we are trying to fix. We detect some edge case (the original reason was to avoid some CME due to reentrant behavior), and then just skip the work we need to do. 

We can't just skip the work that has to be done here, as that leads to incorrectly connected graph states, thus causing the bug we observed.

The second `continue` below was not observed in the original reproducer that drove this PR, but it can likely cause the same behavior (skipping removing an outgoing edge when we were supposed to). In order to resolve _this_ instance, we needed to make additional changes to avoid reentrant behavior specifically in module version selection.